challenges/platform/devops.md

```diff
@@ -0,0 +1,122 @@
+  19: E10FA20A:E6A6 B358BD5B:01BB 01 00000000:00000000 00:00000000 00000000     0        0 26165 1 0000000000000000 28 4 0 10 -1
+  20: E10FA20A:D8B2 9858BD5B:0050 06 00000000:00000000 03:00001219 00000000     0        0 0 3 0000000000000000
+  ```
+
+The second and third columns contain IP addresses and port pairs in hex format. Be careful since the IP address is printed in little endian, and the port is printed in big endian (on x86 architectures).
+
+0664A8C0:1FFE translates to 10.162.15.225:55474
+
+## Level 1:
+
+### Description
+Write a program that reads `/proc/net/tcp` every 10 seconds, and reports any new connections. 
+
+Sample Output:
+```
+2021-04-28 15:28:05: New connection: 192.0.2.56:5973 -> 10.0.0.5:80
+2021-04-28 15:28:05: New connection: 203.0.113.105:31313 -> 10.0.0.5:80
+2021-04-28 15:28:15: New connection: 203.0.113.94:9208 -> 10.0.0.5:80
+2021-04-28 15:28:15: New connection: 198.51.100.245:14201 -> 10.0.0.5:80
+```
+
+Include a readme with the program that explains any dependencies and how to build and execute the program. The interview panel will build and test the program.
+
+### Questions
+1. How would you prove the code is correct?
+2. How would you make this solution better?
+3. Is it possible for this program to miss a connection?
+4. If you weren't following these requirements, how would you solve the problem of logging every new connection? 
+
+## Level 2:
+Implement all of the level 1 requirements plus:
+1. Add a Makefile or your preferred build scripting to build and test the solution.
+2. Add some tests, test for corner cases or unexpected behaviour.
+3. Add the ability to detect a port scan, where a single source IP connects to more than 3 host ports in the previous minute.
+
+Sample Output:
+```
+2021-04-28 15:28:05: Port scan detected: 192.0.2.56 -> 10.0.0.5 on ports 80,81,82,83
+```
+
+### Questions
+1. Why did you choose `x` to write the build automation?
+2. Is there anything else you would test if you had more time?
+3. What is the most important tool, script, or technique you have for solving problems in production? Explain why this tool/script/technique is the most important.
+
+
+## Level 3:
+Implement all of the level 2 requirements plus:
+1. Add a prometheus endpoint to report metrics on the following:
+    1. Counter - number of new connections
+
+    Tip: There are client libraries available in many languages for providing prometheus metrics, see https://prometheus.io/docs/instrumenting/clientlibs/
+2. Build the project into a docker container, provide instructions on how to execute as a container while reporting connections on the host.
+3. When a port scan is detected, configure the host firewall to block connections by source-ip.
+
+### Questions
+1. If you had to deploy this program to hundreds of servers, what would be your preferred method? Why?
+4. What is the hardest technical problem or outage you've had to solve in your career? Explain what made it so difficult?
+
+# Scoring
```

Comment:
@knisbet forgot one thing - we need a scorecard as well. We will add the scorecard in Lever and here. It will guide the interview team and the candidate and tell them how exactly are we going to evaluate their submission.

Reply:
I think this actually got removed from systems engineer since no one was really using it.

Reply:
Talk about perception vs reality :) I totally missed that. I was under impression that it was very helpful for candidates to see how do we rank candidates submission and what areas we are evaluating. If not the scorecard, how do we tell them what aspects of the code/communication we are looking and not looking at?

Reply:
Tried to expand a bit on scoring in 663f3ed21ac7bf10a1a6c6e6e3fa31dca6404f1a. Instead of using a matrix like we used to use, a bit more high level with key focus areas.